File: test/warnings/array_types_warning/cpp/PackedArrayChoiceHasNoPackableFieldTest.cpp
#include "gtest/gtest.h"

#include "array_types_warning/packed_array_choice_has_no_packable_field/PackedArrayChoiceHasNoPackableField.h"

#include "zserio/SerializeUtil.h"

namespace array_types_warning
{
namespace packed_array_choice_has_no_packable_field
{

class PackedArrayChoiceHasNoPackableFieldTest : public ::testing::Test
{
protected:
    static const std::string BLOB_NAME;
};

const std::string PackedArrayChoiceHasNoPackableFieldTest::BLOB_NAME =
        "warnings/array_types_warning/packed_array_choice_has_no_packable_field.blob";

TEST_F(PackedArrayChoiceHasNoPackableFieldTest, writeRead)
{
    PackedArrayChoiceHasNoPackableField packedArrayChoiceHasNoPackableField;

    packedArrayChoiceHasNoPackableField.setArray1({{
        StructWithPackable("A", 65),
        StructWithPackable("B", 66),
        StructWithPackable("C", 67)
    }});

    auto& array2 = packedArrayChoiceHasNoPackableField.getArray2();
    array2.resize(3);
    array2[0].setField2(TestEnum::ONE);
    array2[1].setField2(TestEnum::TWO);
    array2[2].setField2(TestEnum::ONE);

    packedArrayChoiceHasNoPackableField.setArray3({{
        StructWithPackableArray("ABC", {{ 65, 66, 67 }}),
        StructWithPackableArray("DEF", {{ 68, 69, 70 }}),
        StructWithPackableArray("GHI", {{ 71, 72, 73 }})
    }});

    auto& array4 = packedArrayChoiceHasNoPackableField.getArray4();
    array4.resize(3);
    array4[0].initialize(true);
    array4[0].setField1(4.0f);
    array4[1].initialize(true);
    array4[1].setField1(1.0f);
    array4[2].initialize(true);
    array4[2].setField1(0.0f);

    zserio::serializeToFile(packedArrayChoiceHasNoPackableField, BLOB_NAME);
    auto readPackedArrayChoiceHasNoPackableField =
            zserio::deserializeFromFile<PackedArrayChoiceHasNoPackableField>(BLOB_NAME);
    ASSERT_EQ(packedArrayChoiceHasNoPackableField, readPackedArrayChoiceHasNoPackableField);
}

} // namespace packed_array_choice_has_no_packable_field
} // namespace array_types_warning

File: test/warnings/array_types_warning/python/PackedArrayChoiceHasNoPackableFieldTest.py
import os
import zserio

import ArrayTypesWarning

from testutils import getApiDir

class PackedArrayChoiceHasNoPackableFieldTest(ArrayTypesWarning.TestCase):
    def testWriteRead(self):
        packedArrayChoiceHasNoPackableField = self.api.PackedArrayChoiceHasNoPackableField(
            [
                self.api.StructWithPackable("A", 65),
                self.api.StructWithPackable("B", 66),
                self.api.StructWithPackable("C", 67)
            ],
            [
                self.api.UnionWithPackableField(),
                self.api.UnionWithPackableField(),
                self.api.UnionWithPackableField()
            ],
            [
                self.api.StructWithPackableArray("ABC", [65, 66, 67]),
                self.api.StructWithPackableArray("DEF", [68, 69, 70]),
                self.api.StructWithPackableArray("GHI", [71, 72, 73])
            ],
            [
                self.api.ChoiceWithoutPackableField(True),
                self.api.ChoiceWithoutPackableField(True),
                self.api.ChoiceWithoutPackableField(True)
            ]
        )

        packedArrayChoiceHasNoPackableField.array2[0].field2 = self.api.TestEnum.ONE
        packedArrayChoiceHasNoPackableField.array2[1].field2 = self.api.TestEnum.TWO
        packedArrayChoiceHasNoPackableField.array2[2].field2 = self.api.TestEnum.ONE

        packedArrayChoiceHasNoPackableField.array4[0].field1 = 4.0
        packedArrayChoiceHasNoPackableField.array4[1].field1 = 1.0
        packedArrayChoiceHasNoPackableField.array4[2].field1 = 0.0

        zserio.serialize_to_file(packedArrayChoiceHasNoPackableField, self.BLOB_NAME)
        readPackedArrayChoiceHasNoPackableField = zserio.deserialize_from_file(
            self.api.PackedArrayChoiceHasNoPackableField, self.BLOB_NAME
        )
        self.assertEqual(packedArrayChoiceHasNoPackableField, readPackedArrayChoiceHasNoPackableField)

    BLOB_NAME = os.path.join(getApiDir(os.path.dirname(__file__)),
                             "packed_array_choice_has_no_packable_field.blob")
